sktime/forecasting/compose/_skforecast_reduce.py

```diff
@@ -0,0 +1,469 @@
+
+        y_new = y
+        X_new = X
+        # skforecast does not support PeriodIndex and Integer Index.
+        # So converting to supported index types here.
+        if not isinstance(y.index, (pd.RangeIndex, pd.DatetimeIndex)):
+            if isinstance(y.index, pd.PeriodIndex):
+                y_new = self._coerce_period_to_datetime_index(y)
+                if X is not None:
+                    X_new = self._coerce_period_to_datetime_index(X)
+            elif pd.api.types.is_integer_dtype(y.index):
+                y_new = self._coerce_int_to_range_index(y)
+                if X is not None:
+                    X_new = self._coerce_int_to_range_index(X)
+            else:
+                raise ValueError(
+                    f"""y and X must have one of the following index types:
+                    pd.RangeIndex, pd.DatetimeIndex, pd.PeriodIndex, pd.Index
+                    (int dtype). Found index of type: {type(y.index)}"""
+                )
+
+        self._forecaster.fit(y_new, exog=self._coerce_column_names(X_new))
+
+        return self
+
+    def _get_horizon_details(
+        self: "SkforecastAutoreg", fh: Optional[ForecastingHorizon]
+    ):
+        if not fh.is_all_out_of_sample(self.cutoff):
+            raise NotImplementedError(
+                f"{self.__class__.__name__} does not support in-sample predictions."
+            )
+
+        if not (out_of_sample_horizon := fh.to_out_of_sample(self.cutoff)):
+            raise ValueError(
+                f"{self.__class__.__name__} received empty out-of-sample horizon."
+            )
+
+        maximum_forecast_horizon = out_of_sample_horizon.to_relative(self.cutoff)[-1]
+
+        absolute_horizons = out_of_sample_horizon.to_absolute_index(self.cutoff)
+        horizon_positions = out_of_sample_horizon.to_indexer(self.cutoff)
+
+        return maximum_forecast_horizon, absolute_horizons, horizon_positions
+
+    def _predict(
+        self: "SkforecastAutoreg",
+        fh: Optional[ForecastingHorizon],
+        X: Optional[pd.DataFrame],
+    ):
+        """Forecast time series at future horizon.
+
+        private _predict containing the core logic, called from predict
+
+        State required:
+            Requires state to be "fitted".
+
+        Accesses in self:
+            Fitted model attributes ending in "_"
+            self.cutoff
+
+        Parameters
+        ----------
+        fh : guaranteed to be ForecastingHorizon or None, optional (default=None)
+            The forecasting horizon with the steps ahead to to predict.
+        X : pd.DataFrame, optional (default=None)
+            Exogenous time series
+
+        Returns
+        -------
+        y_pred : pd.DataFrame
+            Point predictions
+        """
+        (
+            maximum_forecast_horizon,
+            absolute_horizons,
+            horizon_positions,
+        ) = self._get_horizon_details(fh)
+
+        X_new = X
+        if X is not None:
```

Comment:
The following check seems repetitive, can you address that please?

Reply:
This check is necessary here since X is optional and could be None.

Reply:
I did not mean just L325. These three groups, L246-L264, L324-L336 and L401-L413 are very similar, and last two are probably identical.

Reply:
I can make a common function for them and call it in fit, predict and predict_quantile. Will that be fine?

Reply:
Yes, let's have that as private method.